src/metabase/search/filter.clj

```diff
@@ -0,0 +1,109 @@
+(ns metabase.search.filter
+  (:require
+   [honey.sql.helpers :as sql.helpers]
+   [metabase.driver.common.parameters.dates :as params.dates]
+   [metabase.public-settings.premium-features :as premium-features]
+   [metabase.search.in-place.filter :as search.in-place.filter]
+   [metabase.search.spec :as search.spec]
+   [metabase.util :as u]
+   [metabase.util.date-2 :as u.date]
+   [metabase.util.i18n :refer [tru]])
+  (:import
+   (java.time LocalDate)))
+
+(def ^:private context-key->attr
+  {:archived?           :archived
+   :created-at          :created-at
+   :created-by          :creator-id
+   :last-edited-at      :last-edited-at
+   :last-edited-by      :last-editor-id
+   :search-native-query :dataset-query
+   :verified            :verified})
+
+(def ^:private attr->index-key
+  (into {} (for [k (vals context-key->attr)]
+             [k (keyword (str "search_index." (u/->snake_case_en (name k))))])))
+
+(defn search-context->applicable-models
+  "Returns a set of models that are applicable given the search context.
+
+  If the context has optional filters, the models will be restricted for the set of supported models only."
+  [search-context]
+  (if (= :search-engine/in-place (:search-engine search-context))
+    (search.in-place.filter/search-context->applicable-models search-context)
+    (let [required (keep context-key->attr (keys search-context))]
+      (into #{}
+            (remove nil?)
+            (for [search-model (:models search-context)
+                  :let [spec (search.spec/spec search-model)]]
+              (when (every? (:attrs spec) required)
+                (:name spec)))))))
+
+(defn- date-range-filter-clause
```

Comment:
I didn't bother DRYing this up, since hopefully we can delete the other copy of the code quite soon.